[flang][HLFIR] Relax verifiers of intrinsic operations

tblah opened the pull request:
The verifiers are currently very strict: requiring intrinsic operations to be used only in cases where the Fortran standard permits the intrinsic to be used.

There have now been a lot of cases where these verifiers have caused bugs in corner cases. In a recent ticket, @jeanPerier pointed out that it could be useful for future optimizations if somewhat invalid uses of these operations could be allowed in dead code. See this comment: https://github.com/llvm/llvm-project/issues/79995#issuecomment-1918118234

In response to all of this, I have decided to relax the intrinsic operation verifiers. The intention is now to only disallow operation uses that are likely to crash the compiler. Other checks are still available under `-strict-intrinsic-verifier`.

The disadvantage of this approach is that IR can now represent intrinsic invocations which are incorrect. The lowering and implementation of these intrinsic functions is unlikely to do the right thing in all of these cases, and as they should mostly be impossible to generate using normal Fortran code, these edge cases will see very little testing, before some new optimization causes them to become more common.

Fixes #79995

tblah committed:
[flang][HLFIR] Relax verifiers of intrinsic operations

The verifiers are currently very strict: requiring intrinsic operations
to be used only in cases where the Fortran standard permits the
intrinsic to be used.

There have now been a lot of cases where these verifiers have caused
bugs in corner cases. In a recent ticket, @jeanPerier pointed out that
it could be useful for future optimizations if somewhat invalid uses of
these operations could be allowed in dead code. See this comment:
https://github.com/llvm/llvm-project/issues/79995#issuecomment-1918118234

In response to all of this, I have decided to relax the intrinsic
operation verifiers. The intention is now to only disallow operation
uses that are likely to crash the compiler. It isn't obvious which
checks are more useful than others, and it would not be a good use of time
to try lots of illegal (according to Fortran) intrinsic uses to figure out
exactly which can crash Flang (as none of these should pass semantics
anyway). I have taken an educated guess.

The disadvantage of this approach is that IR can now represent intrinsic
invocations which are incorrect. The lowering and implementation of
these intrinsic functions is unlikely to do the right thing in all of
these cases, and as they should mostly be impossible to generate using
normal Fortran code, these edge cases will see very little testing,
before some new optimization causes them to become more common.

Fixes #79995

jeanPerier commented:
Thanks a lot Tom for applying the comment. I would still keep parts of those checks because we need to enforce some invariants at least to generate codes that would not itself trigger other FIR / LLVM IR verifiers (my comment was maybe a bit too anti-verifiers, sorry for that).

The rank and "relaxed" element types matters to generate addressing and numerical operations that will not throw verifier errors. The length parameters and extents not so much (codegen just need to pick one variable to get the extents/length param and generate code with it).

Looking at these verifiers, I think we can try just relaxing the  "resultExpr.getEleTy() != numTy" to ignore length parameters for the character case, and the extent check is only emitting a warning (this I am in favor or removing altogether or to make it debug only).

Maybe it will belong to type/length propagation to preserve/add some fir.convert when the propagated type would conflict with the original one (i.e. replacing a `?` by anything is OK, but replacing `n` by `m` in in fir.array<> or fir.character<> should not be done).

tblah committed:
Don't throw out as many checks, keep others under an option

This is mostly a revert of the initial commit. See the diff from both
for the intended changes.

jeanPerier reviewed (approved):
Looks great, thanks
